### DISCUSSION TITLE
Is there a simple tool that can send DoH requests to AdGuard Home along with a custom header?

### DISCUSSION THREAD

**Discussion by nmpu** [AdguardTeam/AdGuardHome#6962] [Q&A]
The following **curl** command works for the public **Cloudflare** servers:

`curl -H "accept: application/dns-json" "https://cloudflare-dns.com/dns-query?name=example.com&type=A"`

Is there something similar I can use to test **AdGuard Home**? I've got [https_dns_proxy](https://github.com/aarond10/https_dns_proxy) talking to AGH and I can generate a **ClientID** from the URL suffix. I want to see if I can also pass in an **X-Real-IP** header.

[**trusted_proxies**  If a DNS-over-HTTPS request comes from one of these addresses or networks, AdGuard Home uses the provided proxy headers, such as X-Real-IP, to get the real IP address of the client... ]

**Comment by ghost** [comment]
So long as the proxy's IP is listed in the `trusted_proxies` section of the yaml and the `X-Real-IP` is being forwarded from the proxy correctly, something like the following should work:
`$ curl -H "accept: application/dns-json" -H "X-Real-IP: <client-ip>" "https://<your-adguard-home-instance>/dns-query?name=example.com&type=A"`

#5768 might prove to have valuable information if that doesn't work.

**Reply by nmpu** [reply]
I understand the curl syntax-- at least sending headers. The original curl query does not work on either the AdGuard public server (https://dns.adguard-dns.com/dns-query) or my local instance of AGH. They must not support the GET format like Cloudflare.

The bug report you linked references nginx. I hadn't considered using nginx for the DNS to DoH translation. I think that will work if I can find the correct config. There appears to be some JavaScript translation logic required.

**Comment by andylau004** [comment]
Apifox   or    postman maybe work it

**Comment by nmpu** [comment]
I was able to use the examples [here](https://datatracker.ietf.org/doc/html/rfc8484). You have to manually encode/decode the request/response. In AdGuard Home, the client is displayed as the **ClientID** portion of the request url (/dns-query/**ClientID**/dns=). If you click on the ClientID, you can see the actual client hostname and IP (X-Real-IP works as override). Functionally, it serves my purpose. Visually, I would prefer to see the hostname and IP in the primary view.

**Reply by AAm-kun** [reply]
can you please give a specific example ?
